### ISSUE TITLE
New Sockets API

### ISSUE THREAD

**Issue by fatcerberus** [spheredev/neosphere#25]
Just posting this to document the improvements to the sockets API ahead of the 1.1 release.

There are 3 socket object types in minisphere.
- `Socket` is a Sphere 1.5 legacy socket.  This is what is returned by OpenAddress() and ListenOnPort().  Listening sockets of this type automagically convert to full I/O sockets with the first connection, meaning a new listener must be created afterwards if you intend to run a server.  This type is kept only for backwards compatibility and should not be used in new games.
- `ListeningSocket` is a persistent listener which spawns new I/O sockets for each attempted connection, retrieved one at a time with the `acceptNext()` method.  The `ListeningSocket` maintains a backlog of connections (default size 16); if the backlog fills faster than sockets can be accepted, new connections will begin to be dropped.
- `IOSocket` is exactly what it says on the tin: a socket representing an open connection which can be read from and written to.

The implementation of this API carries with it the extension designation `sphere-new-sockets`.  Extensions can be tested for in script by calling `GetExtensions()`.

**Comment by FlyingJester** [comment]
A more accurate extension would be `minisphere-new-sockets`. Since both TurboSphere and minisphere have similar but inconsistent APIs, both extensions should be namespaced until the APIs are unified.

**Comment by fatcerberus** [comment]
Yeah, I guess I was being a bit presumptuous tagging it as a Sphere extension.

The thing is, I'm not sure how much sense vendor-specific extensions make in a monolithic engine like minisphere.  In TS it's useful because the plugins used can expose different features, but with minisphere I think the single `minisphere` extension it already reports is sufficient.

**Comment by FlyingJester** [comment]
It would mostly be for the games themselves, so that they can check if they are in minisphere, sphere-sfml, or TurboSphere. Since we have different APIs, the game would check which constructor name to use (IOSocket or Socket).

Similarly, if they were checking for something that one engine has that another one doesn't but later gets, they could use the feature immediately in that other engine.

**Comment by fatcerberus** [comment]
True, and if minisphere later gets plugins (which I've been entertaining thoughts of recently), we'd want the differentiated extensions: a single engine-specific extension really isn't enough.

As far as unifying, I'd be willing to compromise a bit by renaming the `acceptNext` method to `accept`.  This would be consistent with BSD sockets, so it's not a huge deal.  Mainly I wanted to communicate that it only accepts one socket at a time, hence I included "next" in the name to highlight that.  Self-documenting APIs are important to me.

The reason I reserved `Socket` for legacy sockets is because, in Sphere, calling toString() on a socket gives "[object socket]", so I'd think most people will expect `new Socket()` to yield the legacy version, just as `new ByteArray()` doesn't give you a Uint8Array back (okay, so it actually does in TS but that's beside the point :-p)
